Properly type CachedFunction & rename to CachedMethod

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114160
* #114162
* __->__ #114161

Previously, I was unsure how to properly type the parameters of a decorated method.
Then I found https://github.com/python/mypy/issues/13222#issuecomment-1193073470
which explains how to use `Concatenate` to hackily achieve it. Not entirely sure why
we can't write a user-defined version of `Callable` that works seamlessly for both functions
and methods...

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

int3 committed:
Properly type CachedFunction & rename to CachedMethod

[ghstack-poisoned]

int3 committed:
Update on "Properly type CachedFunction & rename to CachedMethod"


Previously, I was unsure how to properly type the parameters of a decorated method.
Then I found https://github.com/python/mypy/issues/13222#issuecomment-1193073470
which explains how to use `Concatenate` to hackily achieve it. Not entirely sure why
we can't write a user-defined version of `Callable` that works seamlessly for both functions
and methods...

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 committed:
Update on "Properly type CachedFunction & rename to CachedMethod"


Previously, I was unsure how to properly type the parameters of a decorated method.
Then I found https://github.com/python/mypy/issues/13222#issuecomment-1193073470
which explains how to use `Concatenate` to hackily achieve it. Not entirely sure why
we can't write a user-defined version of `Callable` that works seamlessly for both functions
and methods...

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 committed:
Update on "Properly type CachedFunction & rename to CachedMethod"


Previously, I was unsure how to properly type the parameters of a decorated method.
Then I found https://github.com/python/mypy/issues/13222#issuecomment-1193073470
which explains how to use `Concatenate` to hackily achieve it. Not entirely sure why
we can't write a user-defined version of `Callable` that works seamlessly for both functions
and methods...

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented on torch/_inductor/codegen/memory_planning.py:
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations
```
I had to import this because older Python runtimes don't seem happy when encountering `CachedMethod[[], Any]`; they don't like `[]` being passed as the value of a `ParamSpec` because they treat it as a regular `Generic` typevar. This import makes it so that type hints don't get automatically evaluated at runtime.

This import then made `ruff` un-quote the string type hints in the rest of the file, hence the numerous changes

Skylion007 replied:
@int3 Just want to confirm, using the backported version of typing_extensions.Protocol: https://typing-extensions.readthedocs.io/en/latest/#Protocol does not solve the issue. (Although it appears like it introduces other ones, sadly).

int3 replied:
Hm yeah I don't see how `Protocol` would solve things here...

int3 committed:
Update on "Properly type CachedFunction & rename to CachedMethod"


Previously, I was unsure how to properly type the parameters of a decorated method.
Then I found https://github.com/python/mypy/issues/13222#issuecomment-1193073470
which explains how to use `Concatenate` to hackily achieve it. Not entirely sure why
we can't write a user-defined version of `Callable` that works seamlessly for both functions
and methods...

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented:
@pytorchbot merge
